agreement/asyncVoteVerifier_test.go

```diff
@@ -51,3 +59,466 @@ func TestVerificationAgainstFullExecutionPool(t *testing.T) {
+			sendReceiveVoteVerifications(false, errProb, 0, b.N, b)
+		})
+		if errProb > float32(0.0) {
+			b.Run(fmt.Sprintf("errProb_%.3f_sig_err_only", errProb), func(b *testing.B) {
+				sendReceiveVoteVerifications(true, errProb, 0, b.N, b)
+			})
+		}
+	}
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func sendReceiveVoteVerifications(badSigOnly bool, errProb float32, count, eqCount int, tb testing.TB) {
+	if count+eqCount < 10 {
+		return
+	}
+	voteVerifier := MakeAsyncVoteVerifier(nil)
+	defer voteVerifier.Quit()
+
+	outChan := make(chan asyncVerifyVoteResponse, voteVerifier.Parallelism())
+	gCount := min(200, count)
+	gEqCount := min(200, eqCount)
+
+	errChan := make(chan error)
+	ledger, votes, eqVotes, errsV, errsEqv := generateTestVotes(badSigOnly, errChan, gCount, gEqCount, errProb)
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	if b, ok := tb.(*testing.B); ok {
+		b.ResetTimer()
+	}
+	// collect the verification results and check against the error expectation
+	go func() {
+		defer wg.Done()
+		c := 0
+		for res := range outChan {
+			c++
+			var expectedError error
+			if res.req.uv != nil {
+				expectedError = errsV[int(res.index)]
+			} else {
+				expectedError = errsEqv[int(res.index)]
+			}
+			if (expectedError == nil && res.err != nil) || (expectedError != nil && res.err == nil) {
+				errChan <- fmt.Errorf("expected %v got %v", expectedError, res.err)
+			}
+			if c == count+eqCount {
+				break
+			}
+		}
+		close(errChan)
+	}()
+	// stream the votes to the verifier
+	go func() {
+		defer wg.Done()
+		vi := 0
+		evi := 0
+		for c := 0; c < count+eqCount; c++ {
+			// pick a vote if there are votes, and if either there are no eqVotes or the relative prob
+			turnVote := len(votes) > 0 && (len(eqVotes) == 0 || rand.Float32() < (float32(count)/float32(count+eqCount)))
+			if turnVote {
+				uv := votes[vi%gCount]
+				vi++
+				voteVerifier.verifyVote(context.Background(), ledger, *uv.uv, uint64(uv.id), message{}, outChan)
+			} else {
+				uev := eqVotes[evi%gEqCount]
+				evi++
+				voteVerifier.verifyEqVote(context.Background(), ledger, *uev.uev, uint64(uev.id), message{}, outChan)
+			}
+		}
+	}()
+	// monitor the errors returned from the various goroutines
+	for err := range errChan {
+		require.NoError(tb, err)
+	}
+	wg.Wait()
+}
+
+type unVoteTest struct {
+	uv  *unauthenticatedVote
+	err error
+	id  int
+}
+
+type unEqVoteTest struct {
+	uev *unauthenticatedEquivocationVote
+	err error
+	id  int
+}
+
+func generateTestVotes(onlyBadSigs bool, errChan chan<- error, count, eqCount int, errProb float32) (ledger Ledger,
+	votes []*unVoteTest, eqVotes []*unEqVoteTest, errsV, errsEqv map[int]error) {
+	votes = make([]*unVoteTest, count)
+	eqVotes = make([]*unEqVoteTest, eqCount)
+	errsV = make(map[int]error)
+	errsEqv = make(map[int]error)
+	wg := sync.WaitGroup{}
+	vg := makeTestVoteGenerator()
+
+	nextErrType := 0
+	for c := 0; c < count; c++ {
+		errType := 99
```

Comment:
make `errType` a enum with few named values ?
```
type errType int
const badSig errType = 0
const noErr errType = 99
```
why does `voteOptions` return 9 and `voteEqOptions` - 10 ? what do these magic numbers mean?

Reply:
but one of them is actually named `notSelected`